web/modules/custom/custom_page/src/Plugin/Block/TableOfContentsOutline.php

```diff
@@ -0,0 +1,368 @@
+   */
+  protected $flattenedMenu;
+
+  /**
+   * The og menu tree.
+   *
+   * @var \Drupal\Core\Menu\MenuLinkTreeElement[]
+   */
+  protected $menuTree;
+
+  /**
+   * The og menu instance.
+   *
+   * @var \Drupal\og_menu\OgMenuInstanceInterface
+   */
+  protected $ogMenuInstance;
+
+  /**
+   * Constructs a new RecommendedContentBlock object.
+   *
+   * @param array $configuration
+   *   A configuration array containing information about the plugin instance.
+   * @param string $plugin_id
+   *   The plugin_id for the plugin instance.
+   * @param string $plugin_definition
+   *   The plugin implementation definition.
+   * @param \Drupal\custom_page\CustomPageOgMenuLinksManagerInterface $og_menu_manager
+   *   The og menu manager for custom pages.
+   * @param \Drupal\Core\Routing\RouteMatchInterface $route_match
+   *   The route provider.
+   * @param \Drupal\Core\Menu\MenuLinkTreeInterface $menu_link_tree
+   *   The menu tree manager service.
+   * @param \Drupal\Core\Menu\MenuActiveTrailInterface $active_trail
+   *   The active trail service.
+   */
+  public function __construct(array $configuration, $plugin_id, $plugin_definition, CustomPageOgMenuLinksManagerInterface $og_menu_manager, RouteMatchInterface $route_match, MenuLinkTreeInterface $menu_link_tree, MenuActiveTrailInterface $active_trail) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+    $this->ogMenuManager = $og_menu_manager;
+    $this->node = $route_match->getParameter('node');
+    $this->menuLinkTree = $menu_link_tree;
+    $this->activeTrail = $active_trail;
+    $this->flattenedMenu = [];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('custom_page.og_menu_links_manager'),
+      $container->get('current_route_match'),
+      $container->get('menu.link_tree'),
+      $container->get('menu.active_trail')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function build(): array {
+    $prev = $this->getPrevElement();
+    $up = $this->getParentElement($this->getMenuTree(), $this->getActiveLink()->getPluginId());
+    $next = $this->getNextElement();
+
+    $links = [];
+    if ($prev) {
+      $links['prev'] = [
+        'url' => $prev->getUrlObject(),
+        'title' => $prev->getTitle(),
+      ];
+    }
+    if ($up) {
+      $links['up'] = [
+        'url' => $up->link->getUrlObject(),
+        'title' => $this->t('Up'),
+      ];
+    }
+    if ($next) {
+      $links['next'] = [
+        'url' => $next->getUrlObject(),
+        'title' => $next->getTitle(),
+      ];
+    }
+
+    $build = [
+      '#theme' => 'links',
+      '#links' => $links,
+    ];
+
+    $og_menu_id = $this->getOgMenuName();
+    $build['#cache']['tags'][] = 'config:system.menu.ogmenu-' . $og_menu_id;
```

Comment:
Dunno why, but if I change the menu structure the cache is not invalidated. Added a test for it.